cmd/virt-apiservice-register/virt-apiservice-register.go

```diff
@@ -0,0 +1,737 @@
+		return fmt.Errorf("Secret '%s' doesn't exist", secretName)
+	}
+	if err != nil {
+		return nil
+	}
+
+	// If needed generate a new certificate and update the secret:
+	glog.Infof("Secret '%s' already exists", secretName)
+	key := secret.Data[corev1.TLSPrivateKeyKey]
+	cert := secret.Data[corev1.TLSCertKey]
+	if key == nil || cert == nil {
+		key, cert, err := requestCert(client)
+		if err != nil {
+			return err
+		}
+		if secret.Data == nil {
+			secret.Data = make(map[string][]byte)
+		}
+		secret.Data["ca.crt"] = ca
+		secret.Data[corev1.TLSPrivateKeyKey] = key
+		secret.Data[corev1.TLSCertKey] = cert
+		_, err = secrets.Update(secret)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// findCA tries to find the certificate of the CA that signs the certificates for API servers.
+// Returns an array of bytes containing the CA certificate in PEM format, or nil if the CA can't be
+// found.
+//
+func findCA(client *kubernetes.Clientset) (ca []byte, err error) {
+	// Get the reference to the configuration maps service:
+	configMaps := client.CoreV1().ConfigMaps("kube-public")
+
+	// Retrieve the details of the configuration map:
+	clusterInfo, err := configMaps.Get("cluster-info", metav1.GetOptions{})
+	if err != nil {
+		return
+	}
+
+	// The CA certificate is stored inside a 'kubeconfig' file that is stored inside the
+	// 'cluster-info' configuration map under the 'kubeconfig' key:
+	text := clusterInfo.Data["kubeconfig"]
+	if text == "" {
+		err = fmt.Errorf("Can't find the 'kubeconfig' data")
+		return
+	}
+	config, err := clientcmd.Load([]byte(text))
+	if err != nil {
+		return
+	}
+	cluster := config.Clusters[""]
+	if cluster == nil {
+		err = fmt.Errorf("Can't find the default cluster in the 'kubeconfig' data")
+		return
+	}
+	ca = cluster.CertificateAuthorityData
+	if ca == nil {
+		err = fmt.Errorf("Can't find CA in the 'kubeconfig' data")
+		return
+	}
+	glog.Infof("Found CA\n%s", ca)
+
+	return
+}
+
+// requestCert creates a CSR, sends it to the server and waits till it is approved or denied. If the
+// request is approved it returns arrays of bytes the private key and the certificate, in PEM
+// format.
+//
+func requestCert(client *kubernetes.Clientset) (key, cert []byte, err error) {
+	// Calculate the CN:
+	cn, hosts := findCN()
+	glog.Infof("The CN for the certificate is '%s'", cn)
+	for _, host := range hosts {
+		glog.Infof("Additional host name for the certificate is '%s'", host)
+	}
+
+	// If the CSR already exists it will be impossible to obtain the private key, as it was lost
+	// with the container that created it. That means that we need to create a new CSR, with a
+	// different name. We will try with an increasing number, till there is no such CSR.
+	try := 0
+	name := cn
+	for {
+		_, err = findCSR(client, name)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				break
+			} else {
+				return
+			}
+		}
+		glog.Warningf("The CSR '%s' already exists, will try again with a different name", name)
```

Comment:
I think we should try and save off the progress of the CSR so we can pick up where we left off instead of creating a new one.